File: core/src/main/java/com/tek/core/exception/TekExceptionHandler.java
package com.tek.core.exception;

import com.tek.core.TekCoreProperties;
import com.tek.core.service.TekErrorService;
import com.tek.core.service.TekMailService;
import lombok.NonNull;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.exception.ExceptionUtils;
import org.springframework.core.annotation.Order;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;
import org.springframework.web.context.request.ServletWebRequest;
import org.springframework.web.context.request.WebRequest;
import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler;

/**
 * Tek Core Exception Handler for all REST calls.
 * <p>
 * This service checks if a mail with errors must be sent via
 * {@link org.springframework.mail.javamail.JavaMailSender}
 *
 * @author MarcoPagan
 */
//TODO fine tune of errors
@RestControllerAdvice
@RequiredArgsConstructor
@Slf4j
public class TekExceptionHandler extends ResponseEntityExceptionHandler {

    @NonNull private final TekErrorService errorService;
    @NonNull private final TekMailService mailService;
    @NonNull private final TekCoreProperties coreProperties;

    @ExceptionHandler(value = Exception.class)
    @Order
    public ResponseEntity<Object> handleGenericException(
        RuntimeException ex, WebRequest request
    ) {
        log.error(ExceptionUtils.getStackTrace(ex));
        sendExceptionAsMail(request, ex);
        return handleExceptionInternal(
            ex,
            errorService.createErrorResponse(request),
            new HttpHeaders(),
            HttpStatus.INTERNAL_SERVER_ERROR,
            request
        );
    }

    /**
     * Utility method to check if we have to produce an error mail
     * with the throwable {@link RuntimeException} as attachment.
     *
     * @author MarcoPagan
     */
    private void sendExceptionAsMail(WebRequest request, RuntimeException ex) {
        if (coreProperties.getMail().isSendErrors()) {
            if (coreProperties.getMail().isRealDelivery()) {
                mailService.sendExceptionMessage((ServletWebRequest) request, ex);
            } else {
                log.warn("Parameter sendError is active but realDelivery is false. Skipping mail sending!");
            }
        }
    }
}

File: core/src/main/kotlin/com/tek/core/exception/TekExceptionHandler.kt
package com.tek.core.exception

import com.tek.core.TekErrorResponse
import com.tek.core.i18n.CoreMessageSource
import com.tek.core.i18n.CoreMessageSource.Companion.messageInternalServerError
import org.apache.commons.lang3.exception.ExceptionUtils
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.context.i18n.LocaleContextHolder
import org.springframework.core.Ordered
import org.springframework.core.annotation.Order
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.http.converter.HttpMessageNotReadableException
import org.springframework.validation.FieldError
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ControllerAdvice
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.context.request.ServletWebRequest
import org.springframework.web.context.request.WebRequest
import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler
import java.io.StringWriter
import kotlin.reflect.jvm.internal.impl.utils.ExceptionUtilsKt

/**
 * [org.springframework.web.bind.annotation.RestController] exception handler
 */
@ControllerAdvice
class TekExceptionHandler(
    private val coreMessageSource: CoreMessageSource
) : ResponseEntityExceptionHandler() {

    private val log: Logger = LoggerFactory.getLogger(TekExceptionHandler::class.java)

    /**
     * Function to give a standard response for a [HttpMessageNotReadableException]
     */
    override fun handleHttpMessageNotReadable(
        ex: HttpMessageNotReadableException,
        headers: HttpHeaders,
        status: HttpStatus,
        request: WebRequest
    ): ResponseEntity<Any> {

        log.warn(ex.message)

        return ResponseEntity(
            TekErrorResponse(HttpStatus.BAD_REQUEST).apply {
                this.errors = mapOf("error" to ex.message)
                this.path = (request as ServletWebRequest).request.servletPath
            },
            HttpStatus.BAD_REQUEST
        )
    }

    /**
     * Function to give a standard response for [MethodArgumentNotValidException]
     */
    override fun handleMethodArgumentNotValid(
        ex: MethodArgumentNotValidException,
        headers: HttpHeaders,
        status: HttpStatus,
        request: WebRequest
    ): ResponseEntity<Any> {

        log.warn(ex.message)

        val httpStatus = HttpStatus.NOT_ACCEPTABLE

        return ResponseEntity(
            TekErrorResponse(httpStatus).apply {
                this.errors = ex.bindingResult.allErrors.associate {
                    (it as FieldError).field to it.defaultMessage
                }
                this.path = (request as ServletWebRequest).request.servletPath
            },
            headers,
            httpStatus
        )
    }

    /**
     * Function to give a standard response for a [TekServiceException]
     */
    @ExceptionHandler(TekServiceException::class)
    fun handleServiceException(
        ex: TekServiceException,
        request: WebRequest
    ): ResponseEntity<TekErrorResponse> {

        log.warn(ex.message)

        return ResponseEntity(
            TekErrorResponse(ex.httpStatus).apply {
                this.errors = mapOf("error" to ex.message)
                this.path = (request as ServletWebRequest).request.servletPath
            },
            ex.httpStatus
        )
    }

    /**
     * Function to give a standard response for a [TekResourceNotFoundException]
     */
    @ExceptionHandler(TekResourceNotFoundException::class)
    fun handleServiceException(
        ex: TekResourceNotFoundException,
        request: WebRequest
    ): ResponseEntity<TekErrorResponse> {

        log.warn(ex.message)

        return ResponseEntity(
            TekErrorResponse(ex.httpStatus).apply {
                this.errors = mapOf("error" to ex.message)
                this.path = (request as ServletWebRequest).request.servletPath
            },
            ex.httpStatus
        )
    }

    /**
     * Function to give a standard response for a [TekValidationException]
     */
    @ExceptionHandler(TekValidationException::class)
    fun handleValidationException(
        ex: TekValidationException,
        request: WebRequest
    ): ResponseEntity<TekErrorResponse> {

        val httpStatus = HttpStatus.NOT_ACCEPTABLE
        return ResponseEntity(
            TekErrorResponse(httpStatus).apply {
                this.errors = ex.errors
                this.path = (request as ServletWebRequest).request.servletPath
            },
            httpStatus
        )
    }

    @ExceptionHandler(Exception::class)
    @Order(Ordered.LOWEST_PRECEDENCE)
    fun handleGenericException(
        ex: Exception,
        request: WebRequest
    ): ResponseEntity<TekErrorResponse> {

        log.error(ExceptionUtils.getStackTrace(ex))

        val error = coreMessageSource.getCoreMessageSource()
            .getMessage(messageInternalServerError, null, LocaleContextHolder.getLocale())

        val httpStatus = HttpStatus.INTERNAL_SERVER_ERROR
        return ResponseEntity(
            TekErrorResponse(httpStatus).apply {
                this.errors = mapOf("error" to error)
                this.path = (request as ServletWebRequest).request.servletPath
            },
            httpStatus
        )
    }
}